File: rust/gr_base/src/sys_info.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysSingleCpuInfo {
    #[serde(default)]
    pub name: String,

    #[serde(default)]
    pub usage: f32,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysCpuInfo {
    #[serde(default)]
    pub usage: f32,

    #[serde(default)]
    pub vendor: String,

    #[serde(default)]
    pub brand: String,

    #[serde(default)]
    pub base_frequency: f32,

    #[serde(default)]
    pub current_frequency: f32,

    #[serde(default)]
    pub max_frequency: f32,

    #[serde(default)]
    pub cpus: Vec<SysSingleCpuInfo>,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysMemInfo {
    #[serde(default)]
    pub total: u64,

    #[serde(default)]
    pub total_gb: u64,

    #[serde(default)]
    pub used: u64,

    #[serde(default)]
    pub used_gb: u64,

    #[serde(default)]
    pub available: u64,

    #[serde(default)]
    pub available_gb: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysDiskInfo {
    #[serde(default)]
    pub disk_type: String, // ssd or ...

    #[serde(default)]
    pub mount_on: String, //C: D:

    #[serde(default)]
    pub filesystem: String,

    #[serde(default)]
    pub available: u64,

    #[serde(default)]
    pub available_gb: u64,

    #[serde(default)]
    pub total: u64,

    #[serde(default)]
    pub total_gb: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysIpNetwork {
    #[serde(default)]
    pub addr: String,

    #[serde(default)]
    pub prefix: u8,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysNetworkInfo {
    #[serde(default)]
    pub name: String,

    #[serde(default)]
    pub mac: String,

    #[serde(default)]
    pub ip_networks: Vec<SysIpNetwork>,

    #[serde(default)]
    pub received_data: u64,

    #[serde(default)]
    pub sent_data: u64,

    #[serde(default)]
    pub max_transmit_speed: u64,

    #[serde(default)]
    pub max_receive_speed: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysUserInfo {}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysOsInfo {
    #[serde(default)]
    pub sys_name: String,

    #[serde(default)]
    pub sys_kernel_version: String,

    #[serde(default)]
    pub sys_os_version: String,

    #[serde(default)]
    pub sys_os_long_version: String,

    #[serde(default)]
    pub sys_host_name: String,

    #[serde(default)]
    pub sys_kernel: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysComponentInfo {
    #[serde(default)]
    pub temperature: f32,

    #[serde(default)]
    pub max: f32,

    #[serde(default)]
    pub critical: f32,

    #[serde(default)]
    pub label: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysGpuInfo {
    #[serde(default)]
    pub id: String,

    #[serde(default)]
    pub brand: String,

    #[serde(default)]
    pub fan_speed: u32,

    #[serde(default)]
    pub power_limit: u32,

    #[serde(default)]
    pub encoder_utilization: u32,

    #[serde(default)]
    pub gpu_utilization: u32,

    #[serde(default)]
    pub mem_utilization: u32,

    #[serde(default)]
    pub temperature: u32,

    #[serde(default)]
    pub mem_free: u64,

    #[serde(default)]
    pub mem_free_gb: f32,

    #[serde(default)]
    pub mem_used: u64,

    #[serde(default)]
    pub mem_used_gb: f32,

    #[serde(default)]
    pub mem_total: u64,

    #[serde(default)]
    pub mem_total_gb: f32,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
pub struct SysInfo {
    #[serde(default)]
    pub timestamp: i64,

    #[serde(default)]
    pub timestamp_readable: String,

    #[serde(default)]
    pub cpu: SysCpuInfo,

    #[serde(default)]
    pub mem: SysMemInfo,

    #[serde(default)]
    pub disks: Vec<SysDiskInfo>,

    #[serde(default)]
    pub networks: Vec<SysNetworkInfo>,

    #[serde(default)]
    pub os: SysOsInfo,

    #[serde(default)]
    pub components: Vec<SysComponentInfo>,

    #[serde(default)]
    pub uptime: String,

    #[serde(default)]
    pub gpus: Vec<SysGpuInfo>,
}

File: web/gr_cms/src/entity/sys_info.ts
/* ================= CPU ================= */

export class SysSingleCpuInfo {
  constructor(
    public name: string = '',
    public usage: number = 0,
  ) {}
}

export class SysCpuInfo {
  constructor(
    public usage: number = 0,
    public vendor: string = '',
    public brand: string = '',
    public base_frequency: number = 0,
    public current_frequency: number = 0,
    public max_frequency: number = 0,
    public cpus: SysSingleCpuInfo[] = [],
  ) {}
}

/* ================= Memory ================= */

export class SysMemInfo {
  constructor(
    public total: number = 0,
    public total_gb: number = 0,
    public used: number = 0,
    public used_gb: number = 0,
    public available: number = 0,
    public available_gb: number = 0,
  ) {}
}

/* ================= Disk ================= */

export class SysDiskInfo {
  constructor(
    public disk_type: string = '',
    public mount_on: string = '',
    public filesystem: string = '',
    public available: number = 0,
    public available_gb: number = 0,
    public total: number = 0,
    public total_gb: number = 0,
  ) {}
}

/* ================= Network ================= */

export class SysIpNetwork {
  constructor(
    public addr: string = '',
    public prefix: number = 0,
  ) {}
}

export class SysNetworkInfo {
  constructor(
    public name: string = '',
    public mac: string = '',
    public ip_networks: SysIpNetwork[] = [],
    public received_data: number = 0,
    public sent_data: number = 0,
    public max_transmit_speed: number = 0,
    public max_receive_speed: number = 0,
  ) {}
}

/* ================= User ================= */

export class SysUserInfo {
  // 目前 Rust 里是空 struct
  constructor() {}
}

/* ================= OS ================= */

export class SysOsInfo {
  constructor(
    public sys_name: string = '',
    public sys_kernel_version: string = '',
    public sys_os_version: string = '',
    public sys_os_long_version: string = '',
    public sys_host_name: string = '',
    public sys_kernel: string = '',
  ) {}
}

/* ================= Component ================= */

export class SysComponentInfo {
  constructor(
    public temperature: number = 0,
    public max: number = 0,
    public critical: number = 0,
    public label: string = '',
  ) {}
}

/* ================= GPU ================= */

export class SysGpuInfo {
  constructor(
    public id: string = '',
    public brand: string = '',
    public fan_speed: number = 0,
    public power_limit: number = 0,
    public encoder_utilization: number = 0,
    public gpu_utilization: number = 0,
    public mem_utilization: number = 0,
    public temperature: number = 0,
    public mem_free: number = 0,
    public mem_free_gb: number = 0,
    public mem_used: number = 0,
    public mem_used_gb: number = 0,
    public mem_total: number = 0,
    public mem_total_gb: number = 0,
  ) {}
}

/* ================= Root ================= */

export class SysInfo {
  constructor(
    public timestamp: number = 0,
    public timestamp_readable: string = '',
    public cpu: SysCpuInfo = new SysCpuInfo(),
    public mem: SysMemInfo = new SysMemInfo(),
    public disks: SysDiskInfo[] = [],
    public networks: SysNetworkInfo[] = [],
    public os: SysOsInfo = new SysOsInfo(),
    public components: SysComponentInfo[] = [],
    public uptime: string = '',
    public gpus: SysGpuInfo[] = [],
  ) {}
}
